test/system/src/main/java/io/pravega/test/system/framework/services/docker/DockerBasedService.java

```diff
@@ -0,0 +1,199 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.test.system.framework.services.docker;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.spotify.docker.client.DefaultDockerClient;
+import com.spotify.docker.client.DockerClient;
+import com.spotify.docker.client.exceptions.DockerException;
+import com.spotify.docker.client.messages.ServiceCreateResponse;
+import com.spotify.docker.client.messages.swarm.EndpointSpec;
+import com.spotify.docker.client.messages.swarm.PortConfig;
+import com.spotify.docker.client.messages.swarm.Service;
+import com.spotify.docker.client.messages.swarm.ServiceMode;
+import com.spotify.docker.client.messages.swarm.ServiceSpec;
+import com.spotify.docker.client.messages.swarm.TaskSpec;
+import com.spotify.docker.client.messages.swarm.TaskStatus;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import static io.pravega.test.system.framework.DockerRemoteSequential.DOCKER_CLIENT_PORT;
+import static org.junit.Assert.assertNotNull;
+import lombok.extern.slf4j.Slf4j;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import com.spotify.docker.client.messages.swarm.Task;
+
+@Slf4j
+public abstract class DockerBasedService implements io.pravega.test.system.framework.services.Service {
+
+    static final int ZKSERVICE_ZKPORT = 2181;
+    static final String IMAGE_PATH = System.getProperty("dockerImageRegistry");
+    static final String PRAVEGA_VERSION = System.getProperty("imageVersion");
+    final DockerClient dockerClient;
+    final String serviceName;
+    private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(3);
```

Comment:
I don't see this shutdown. Perhaps it should be tied to the start/stop lifecycle?

Reply:
its done here, 
```
return CompletableFuture.runAsync(() -> {
            startTest(containerName, className, methodName);
        }).thenCompose(v2 -> waitForJobCompletion())
                .<Void>thenApply(v1 -> {
                    try {
                        if (Exceptions.handleInterrupted(() -> client.inspectContainer(id.get()).state().exitCode() != 0)) {
                            throw new AssertionError("Test failed "
                                    + className + "#" + methodName);
                        }
                        executorService.shutdown();
```